lib/parser/canonicalize-omp.cc

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "canonicalize-omp.h"
+#include "parse-tree-visitor.h"
+#include "parse-tree.h"
+
+// After Loop Canonicalization, rewrite OpenMP parse tree to make OpenMP
+// Constructs more structured which provide explicit scopes for later
+// structural checks and semantic analysis.
+//   1. move structured DoConstruct and OpenMPEndloopdirective into
+//      OpenMPLoopConstruct. Compilation will not proceed in case of errors
+//      after this pass.
+//   2. TODO: Start and End directive pair matching
+//   3. TBD
+namespace Fortran::parser {
+
+class CanonicalizationOfOmp {
+public:
+  template<typename T> bool Pre(T &) { return true; }
+  template<typename T> void Post(T &) {}
+  CanonicalizationOfOmp(Messages &messages) : messages_{messages} {}
+
+  void Post(Block &block) {
+    for (auto it{block.begin()}; it != block.end(); ++it) {
+      if (auto *exec{std::get_if<ExecutableConstruct>(&it->u)}) {
+        if (auto *ompCons{
+                std::get_if<common::Indirection<OpenMPConstruct>>(&exec->u)}) {
+          // OpenMPLoopConstruct
+          if (auto *ompLoop{
+                  std::get_if<OpenMPLoopConstruct>(&ompCons->value().u)}) {
+            RewriteOpenMPLoopConstruct(ompLoop, block, it);
+          }
+        } else if (auto *endDir{
+                       std::get_if<common::Indirection<OpenMPEndLoopDirective>>(
+                           &exec->u)}) {
```

Comment:
You can use `GetConstructIf` to get `ompCons` and `ompLoop` too.

Reply:
I've updated the `ompCons` and `endDir` to use `GetConstructIf`. Sorry for the back & forth.